segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameReader.java

```diff
@@ -137,5 +170,87 @@ public DataFrameRecord<T> getNext() throws DataCorruptionException, DurableDataL
         }
     }
 
+    /**
+     * Adds a new entry to the processed log items tracking map (only in case the log entry is related to a new sequence
+     * number). Also, deletes the oldest element in the map in case its size exceeds maxTrackedEntriesToCheckDuplicates.
+     *
+     * @param seqNo Sequence number of the new log item to track.
+     * @param logItem Internal information for this log item.
+     */
+    private void trackLastReadLogItem(long seqNo, DataFrameRecord.RecordInfo logItem) {
+        if (this.duplicateEntryTracker.containsKey(seqNo)) {
+            log.warn("{}: Found duplicate entry {}. Do not track it.", this.traceObjectId, seqNo);
+            return;
+        }
+
+        // We have a new valid log entry to track, just add it to the tracking map.
+        this.duplicateEntryTracker.put(seqNo, logItem);
+
+        // Remove exceeding elements from tracking map.
+        if (this.duplicateEntryTracker.getNumTrackedElements() > this.maxOverlapToCheckForDuplicates) {
+            this.duplicateEntryTracker.removeLast();
+        }
+    }
+
+    private boolean isHarmlessDuplicateEntry(long seqNo, DataFrameRecord.RecordInfo recordInfo) {
+        // Check if the duplicate comes from too far behind. In this case, we will report DCE anyway.
+        if (!duplicateEntryTracker.containsKey(seqNo)) {
+            log.warn("{}: Seems that we have found a duplicate {}, but it exceeds the max number of tracking elements {}.",
+                    this.traceObjectId, seqNo, this.maxOverlapToCheckForDuplicates);
+            return false;
+        }
+
+        // To determine if duplicate log entry is harmless, we use the available internal information of a DataFrame to
+        // check for equality: i) length of DataFrameRecord.EntryInfo list, and ii) compare that all DataFrameRecord.EntryInfo
+        // have the same length and offset within the DataFrame.
+        boolean harmlessDuplicate = duplicateEntryTracker.getEntriesFor(seqNo).size() == recordInfo.getEntries().size();
```

Comment:
cant we fail-fast here?

Reply:
Well, it will fail anyway as soon as the next `for` completes, as the `harmlessDuplicate` boolean is updated with `&=`. I could add another `if` clause to return `false` directly, but I thought this makes the code simpler.